src/many-identity-webauthn/src/identity.rs

```diff
@@ -0,0 +1,131 @@
+#![cfg(feature = "identity")]
+
+mod u2fhid;
+
+use coset::cbor::value::Value;
+use coset::{iana, CborSerializable, CoseKey, CoseSign1, KeyOperation, Label};
+use many_error::ManyError;
+use many_identity::{Address, Identity};
+use many_identity_dsa::ecdsa;
+use many_modules::idstore;
+use many_protocol::ManyUrl;
+use sha2::{Digest, Sha512};
+use webauthn_authenticator_rs::AuthenticatorBackend;
+use webauthn_rs::prelude::Url;
+use webauthn_rs_proto::PublicKeyCredentialRequestOptions;
+use webauthn_rs_proto::{AllowCredentials, AuthenticatorTransport, UserVerificationPolicy};
+
+pub struct WebAuthnIdentity {
+    address: Address,
+    public_key: CoseKey,
+    cred_id: idstore::CredentialId,
+    origin_url: ManyUrl,
+    rp_id: String,
+}
+
+impl WebAuthnIdentity {
+    pub fn authenticate(
+        origin_url: ManyUrl,
+        rp_id: String,
+        creds: idstore::GetReturns,
+    ) -> Result<Self, ManyError> {
+        let mut public_key = CoseKey::from_slice(creds.public_key.0.as_slice())
+            .map_err(ManyError::deserialization_error)?;
+        public_key
+            .key_ops
+            .insert(KeyOperation::Assigned(iana::KeyOperation::Verify));
+
+        Ok(Self {
+            address: ecdsa::address(&public_key)?,
+            public_key,
+            cred_id: creds.cred_id,
+            origin_url,
+            rp_id,
+        })
+    }
+}
+
+impl Identity for WebAuthnIdentity {
+    fn address(&self) -> Address {
+        self.address
+    }
+
+    fn public_key(&self) -> Option<CoseKey> {
+        Some(self.public_key.clone())
+    }
+
+    fn sign_1(&self, envelope: CoseSign1) -> Result<CoseSign1, ManyError> {
+        let mut envelope = many_identity::cose::add_keyset_header(envelope, self)?;
+
+        envelope
+            .protected
+            .header
+            .rest
+            .push((Label::Text("webauthn".to_string()), Value::Bool(true)));
+        envelope.protected.header.key_id = self.address.to_vec();
+
+        let mut hash = Sha512::new();
+        if let Some(payload) = &envelope.payload {
+            hash.update(payload);
+        }
+
+        let hash = hash.finalize().to_vec();
+
+        let protected_header = envelope.protected.clone();
+        let challenge: Result<Vec<u8>, minicbor::encode::Error<_>> = (|| {
+            let mut encoder = minicbor::Encoder::new(Vec::new());
+            encoder
+                .map(2)?
+                .u8(0)?
```

Comment:
nit: Use a constant. Same with the other one below.

Reply:
Moved to a special type. Simpler.